### ISSUE TITLE
incorrect 'UP-TO-DATE' tasks

### ISSUE THREAD

**Issue by TheBoegl** [TheBoegl/gradle-launch4j#11]
The inputs and outputs of the tasks are wrong which results in 'up-to-date' task which, in reality, are not.

**Comment by TheBoegl** [comment]
Has someone a better idea than e9653f4a?

**Comment by byronpdx** [comment]
I noticed that the copyL4jLib comes back with UP-TO-DATE but doesn't generate any libraries. For the application I am working on I would like it to copy the libraries to the correct location. It could be because I have dome something wrong in the configuration, but would love to know how to make it work. On another project it seems to work fine.

**Comment by TheBoegl** [comment]
Could you please post/link your build script, which shows at least the launch4j configuration, version, changed dependencies and your custom copy task?

**Comment by TheBoegl** [comment]
@byronpdx Just to double check: The copyL4jLib task is not supposed to generate the libraries for you. All it does is fetching and copying the created jar(s) by the jar task and it's dependencies into the launch4j/lib folder. Did you apply the JavaPlugin?

**Comment by byronpdx** [comment]
Yes, I applied the JavaPlugin. Not sure what is supposed to happen with the copyL4jLib task but I am not getting any launch4j/lib folder created nor anything copied into it. Here is my build file
```java
plugins {
    id 'java'
    id 'application'
    id "edu.sc.seis.launch4j" version "1.4.1"
}

sourceCompatibility = 1.8
version = "2.0.2"
mainClassName = "com.byronpdx.launcher.LauncherMain"


repositories {
    mavenCentral()
}


dependencies {
    compile "ch.qos.logback:logback-classic:1.1.3",
            "org.slf4j:log4j-over-slf4j:1.7.10",
            'org.mybatis:mybatis:3.3.0',
            'org.mariadb.jdbc:mariadb-java-client:1.2.0',
            'com.google.code.gson:gson:2.3.1',
            project('NewCommon')

    testCompile 'junit:junit:4.11'
}

// launch4J configuration

copyL4jLib.enabled = true

launch4j {
    outputDir = "launch4j"
    xmlFileName = "launch4j.xml"
    mainClassName = project.mainClassName
    dontWrapJar = true
    headerType = "gui"
    jar = "lib/${project.tasks[JavaPlugin.JAR_TASK_NAME].archiveName}"
    outfile = project.name + '.exe'
    errTitle = ""
    cmdLine = ""
    chdir = '.'
    priority = 'normal'
    downloadUrl = "http://java.com/download"
    supportUrl = ""
    stayAlive = false
    restartOnCrash = false
    manifest = ""
    icon = "$projectDir/launch.ico"
    version = project.version
    textVersion = project.version
    copyright = "Byron Palmer (2015)"
    companyName = "First Group"
    description = project.name
    productName = project.name
    internalName = project.name
    opt = ""
    bundledJre64Bit = false
    bundledJreAsFallback = false
    jreMinVersion = "${project.sourceCompatibility}.0"
    jdkPreference = "preferJre"
    jreRuntimeBits = "64/32"
    messagesStartupError = "Failed to start, run from cmd line"
    messagesBundledJreError = "No suitable JRE found"
    messagesJreVersionError = "Upgrade your java version"
    messagesLauncherError = "Launch fails"
    maxHeapSize = 300
}
```
When I run launch4j my output looks like:
```
Adding the maven central repository to retrieve the launch4j files.
:copyL4jLib UP-TO-DATE
:generateXmlConfig UP-TO-DATE
:createExeWithBin SKIPPED
:unzipL4jBin UP-TO-DATE
:copyL4jBinLib UP-TO-DATE
:createExeWithJar UP-TO-DATE
:createExe UP-TO-DATE
:launch4j UP-TO-DATE
```
What I get created is a folder `launch4j` with a `bin` folder inside, `launch4j.xml` and `Launcher.exe`.

To make this something that can be distributed I run the installDist and then copy the files into a new folder using 
```java
task prepareDist {
    delete '$buildDir/dist/'
    copy {
        from("$buildDir/install/${project.name}/lib")
        into "$buildDir/dist/${project.name}/lib"
    }
    copy {
        from("$buildDir/launch4j/${project.name}.exe")
        into "$buildDir/dist/${project.name}"
    }
    copy {
        from("$projectDir/conf/")
        into "$buildDir/dist/${project.name}/conf"
    }
    copy {
        from("$projectDir/launcher.json")
        into "$buildDir/dist/${project.name}"
    }
    copy {
        from("$projectDir/icons")
        into "$buildDir/dist/${project.name}/icons"
    }
    doLast {
        println "Completed dependency copy"
    }
}
```
I suspect that there are better ways to do this but it seems to work for now.

As I said earlier, I am just learning gradle as well as Intellij and trying to get projects out the door with deadlines so I don't get to dig as deeply and thoroughly as I would like. I appreciate all that you are doing and I see from examples that you have given how to improve what I am doing.

**Comment by TheBoegl** [comment]
First of all: You don't have to initialize the default launch4j properties with their default values.

There seem to be an error on the dependencies if you load this plugin with the plugin-loading mechanism. Could you please try to use the *apply* mechanism to load this plugin and try again if the jar and runtime configuration is loaded correctly and copied to the lib folder?

I'll check that issue when I'm back home from vacation in three weeks.

**Comment by TheBoegl** [comment]
If you apply this plugin instead of the new plugins mechanism it works correctly.
I'll look into this...

**Comment by Georgeto** [comment]
Any updates?

**Comment by TheBoegl** [comment]
A possible fix is already in the develop branch, though I haven't had time for further testing and linking it to this thread.

**Comment by TheBoegl** [comment]
96c30231a8b2866c6ab99667ffd9c0721dfafaf9, acee7fa1ade5fd8d9dbfd1c02e15c2cef49199c8 and 9782a7ed6e88943e2ac443d6a1d9f9f1ab7ce9f0 should fix this issue.

**Comment by thibautd** [comment]
If the fix is available in 1.5.3, I still have the dependency issue when using the plugin loading mechanism instead of 'apply'.

**Comment by TheBoegl** [comment]
@thibautd Please specify which dependency issue you have.
